src/main/java/org/threadly/concurrent/wrapper/limiter/RateLimiterExecutor.java

```diff
@@ -305,26 +304,34 @@ public long execute(double permits, Runnable task) {
   }
   
   private long taskDelayForPermits(double permits) {
-    double effectiveDelay = (permits / permitsPerSecond) * 1000;
-    synchronized (permitLock) {
-      if (permits == 0 && lastScheduleTime < Clock.lastKnownForwardProgressingMillis()) {
-        // shortcut
+    if (permits == 0) { // shortcut since we don't need to update
+      long lastScheduleTimeLong = (long)Double.longBitsToDouble(lastScheduleTime.get());
+      if (lastScheduleTimeLong > Clock.lastKnownForwardProgressingMillis()) {
+        return Math.max(0, lastScheduleTimeLong - Clock.accurateForwardProgressingMillis());
+      } else {
         return 0;
+      }
+    }
+    
+    double effectiveDelay = (permits / permitsPerSecond) * 1000;
+    long now = Clock.accurateForwardProgressingMillis();
+    while (true) {
+      long casLong = lastScheduleTime.get();
+      double lastTimeDouble = Double.longBitsToDouble(casLong);
+      double scheduleDelay = lastTimeDouble - now;
+      if (scheduleDelay > maxScheduleDelayMillis) {
+        return -1 ;  // let rejection handler invoke outside of lock
       } else {
-        double scheduleDelay = lastScheduleTime - Clock.accurateForwardProgressingMillis();
-        if (scheduleDelay > maxScheduleDelayMillis) {
-          // let rejection handler invoke outside of lock
-          return -1;
-        } else if (scheduleDelay < 1) {
-          if (scheduleDelay < 0) {
-            lastScheduleTime = Clock.lastKnownForwardProgressingMillis() + effectiveDelay;
-          } else {
-            lastScheduleTime += effectiveDelay;
-          }
-          return 0;
+        if (scheduleDelay < 0) {
+          if (lastScheduleTime.compareAndSet(casLong, 
+                                             Double.doubleToRawLongBits(now + effectiveDelay))) {
+            return 0;
+          } // else loop and retry
         } else {
-          lastScheduleTime += effectiveDelay;
-          return (long)scheduleDelay;
+          if (lastScheduleTime.compareAndSet(casLong, 
+                                             Double.doubleToRawLongBits(lastTimeDouble + effectiveDelay))) {
+            return (long)scheduleDelay;
```

Comment:
I feel like you could end up with this being "0.9", returning a 0, granted that less then 1ms, not sure if it matters much but you do treat < 0 a different above.